### PR TITLE
Minor updates, CRTP message

### PR DESCRIPTION
**Pull request by three0s**
CRTP:
https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern
I don't think it accomplishes anything. But I am trying it out anyway.

Also some other minor changes.

### CODE REVIEW

File: shared/message.h [hunk 1]
```diff
@@ -22,7 +23,34 @@ enum class cmdtype : std::size_t {
...
+        j.at("success").get_to(success);
+        self().merge_response(j);
+    }
+    void base_from_request(const nlohmann::json &j) { self().from_request(j); }
+
+    bool success = false;
+};
+
+template<typename T>
+using IsMessage = typename std::enable_if<std::is_base_of<base_message<T>, T>::value>::type;
```
Comment (joey-xu) [inline shared/message.h]: Can you explain this statement?

File: shared/message.h [hunk 2]
```diff
@@ -142,23 +163,21 @@ struct message_fusemagic {
...
     // use the same message for equipping & unequipping magics
     const cmdtype type = cmdtype::reordermagic;
 
     message_reordermagic() = default;
-    message_reordermagic(std::vector<int>&& equipped_magic_ids);
+    explicit message_reordermagic(std::vector<int> &&equipped_magic_ids);
     bool validate(const nibashared::sessionstate &session);
     nlohmann::json create_response();
     nlohmann::json create_request();
     void merge_response(const nlohmann::json &j);
```
Comment (joey-xu) [inline shared/message.h]: Since we now use base_* methods? These validate, create and merge are no longer used?

File: shared/message.h (hunk 1 above)
Comment (three0s) [reply shared/message.h]: this checks if `T` is derived from `base_message<T>` which can be used as 
`typename = IsMessage<Message>` in a template

File: shared/message.h (hunk 2 above)
Comment (three0s) [reply shared/message.h]: the `base_*` methods actually call these

Comment (joey-xu) [reply shared/message.h]: I see. Basically, it's the same static polymorphism as introduced in the wiki page. The template parameter is passed into the base class at compile time. There is a Capture Conversion concept in Java which seems to be very similar. Thanks for bringing this interesting feature here.

**Pull request merged**